File: app/Sources/Parley/GoogleTTS.swift
import Foundation

// Google Cloud Text-to-Speech (Chirp3 HD). Cheap (~free under 1M chars/month), good
// German neural voices. One-shot synthesize endpoint → base64 LINEAR16 (WAV) → PCM.
// API key sent via the X-Goog-Api-Key header (never in the URL).
enum GoogleTTS {
    static let endpoint = "https://texttospeech.googleapis.com/v1/text:synthesize"
    static let defaultVoice = "de-DE-Chirp3-HD-Alnilam"
    static let sampleRate = 24000

    static func request(text: String, apiKey: String, voice: String) -> URLRequest {
        var req = URLRequest(url: URL(string: endpoint)!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(apiKey, forHTTPHeaderField: "X-Goog-Api-Key")
        // languageCode = first two segments of the voice name, e.g. "de-DE".
        let lang = voice.split(separator: "-").prefix(2).joined(separator: "-")
        let body: [String: Any] = [
            "input": ["text": text],
            "voice": ["languageCode": lang, "name": voice],
            "audioConfig": ["audioEncoding": "LINEAR16", "sampleRateHertz": sampleRate],
        ]
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)
        return req
    }

    /// Extract raw 16-bit LE PCM from the JSON response (strips the WAV header if present).
    static func pcm(from data: Data) -> Data? {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let b64 = obj["audioContent"] as? String,
              let audio = Data(base64Encoded: b64) else { return nil }
        if audio.count > 44, audio.prefix(4) == Data("RIFF".utf8) {
            return audio.subdata(in: 44..<audio.count)
        }
        return audio
    }
}

// Learned synthesis-latency model: EWMA of seconds-per-character, persisted across runs.
// Used to pause media just ~1s BEFORE the TTS is predicted to be ready instead of at turn
// start — YouTube keeps playing during most of the synthesis wait.
enum TTSTiming {
    private static let key = "parley.ttsSecPerChar"

    /// Predicted synthesis duration for a text of `chars` characters (clamped 0.3–8 s).
    static func predict(chars: Int) -> Double {
        let stored = UserDefaults.standard.double(forKey: key)
        let secPerChar = stored > 0 ? stored : 0.012   // ~3s for a 250-char line, pre-learning
        return min(8, max(0.3, secPerChar * Double(max(chars, 1))))
    }

    /// Feed an observed synthesis (chars → seconds); EWMA alpha 0.3.
    static func record(chars: Int, seconds: Double) {
        guard chars > 0, seconds > 0.05, seconds < 30 else { return }
        let sample = seconds / Double(chars)
        let old = UserDefaults.standard.double(forKey: key)
        UserDefaults.standard.set(old > 0 ? old * 0.7 + sample * 0.3 : sample, forKey: key)
    }
}

File: windows/GoogleTts.cs
using System.Text;
using System.Text.Json;

namespace Parley;

// Google Cloud Text-to-Speech (Chirp3 HD) — identical request shape to the macOS app.
// Returns raw 16-bit LE mono PCM at 24 kHz (WAV header stripped).
public static class GoogleTts
{
    public const int SampleRate = 24000;
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public static async Task<byte[]?> Synthesize(string text, Config config)
    {
        if (!config.UseGoogle) return null;
        var lang = string.Join("-", config.GoogleVoice.Split('-').Take(2));
        var payload = new
        {
            input = new { text },
            voice = new { languageCode = lang, name = config.GoogleVoice },
            audioConfig = new { audioEncoding = "LINEAR16", sampleRateHertz = SampleRate },
        };
        using var req = new HttpRequestMessage(HttpMethod.Post,
            "https://texttospeech.googleapis.com/v1/text:synthesize");
        req.Headers.Add("X-Goog-Api-Key", config.GoogleKey);
        req.Content = new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");
        try
        {
            using var resp = await Http.SendAsync(req);
            var body = await resp.Content.ReadAsStringAsync();
            if (!resp.IsSuccessStatusCode)
            {
                Log.Write($"google tts http {(int)resp.StatusCode}: {body[..Math.Min(200, body.Length)]}");
                return null;
            }
            using var doc = JsonDocument.Parse(body);
            if (!doc.RootElement.TryGetProperty("audioContent", out var b64)) return null;
            var audio = Convert.FromBase64String(b64.GetString() ?? "");
            // Strip the 44-byte WAV header if present.
            if (audio.Length > 44 && audio[0] == (byte)'R' && audio[1] == (byte)'I' &&
                audio[2] == (byte)'F' && audio[3] == (byte)'F')
                return audio[44..];
            return audio;
        }
        catch (Exception e)
        {
            Log.Write($"google tts error: {e.Message}");
            return null;
        }
    }
}
